extension/src/completion/autocompletionProvider.ts

```diff
@@ -171,30 +187,222 @@ function getMatchingWord(document: vscode.TextDocument, position: vscode.Positio
+                    context = 2; // this context tells me I am looking for a variable declaration
+                    this.tryAddSymbol(atom.symbol, AutoLispKind.Global);
+                } else {
+                    this.previousItems.unshift(ucAtom);
+                }
+            }
+        });
+    }
+
+    load(document: vscode.TextDocument, position: vscode.Position): void {
+        // get the scope of this position  
+        let kwRegex: RegExp = /\(\s*(DEFUN|DEFUN-Q)\s+\S{1,99}(\s|\(|;)/gi;
+        let tmp = findContainers(document, position);
+        let pr1: ElementRange = tmp.containerBlockComment;
+        let pr2: ElementRange[] = tmp.containerParens;
+        let contents: string = "";
+        const containers: ElementRange[] = findContainers(document, position).containerParens;
+        for (let ci = 0; ci < containers.length; ci++) {
+            const entry: ElementRange = containers[ci];
+            const value: string = document.getText(new vscode.Range(document.positionAt(entry.startPos.offsetInDocument), document.positionAt(entry.endPos.offsetInDocument)));
+            // retrieve the string contents of the first Defun scope
+            if (value.search(kwRegex) !== -1){                
+                contents = value;
+                break;
+            }
+        }
+        // Fallback for when the typed values are not in a Defun
+        if (contents === ""){
+            contents = document.getText(new vscode.Range(document.positionAt(containers.slice(-1)[0].startPos.offsetInDocument), document.positionAt(containers.slice(-1)[0].endPos.offsetInDocument)));
+        }
+        
+        
+        // Create and load that scope into a temporary document for contextual parsing, had to add this openFromContent method to narrow the LispParser SOW
+        const tempDoc: ReadonlyDocument = ReadonlyDocument.openFromContent(contents, "autolisp");
+        const lp: LispParser = new LispParser(tempDoc);        
+        lp.tokenizeString(contents, 0); 
+        // this atoms forest should only be comprised of a single Sexpression unless
+        if (lp.atomsForest.length === 1 && lp.atomsForest[0] instanceof Sexpression){
+            this.traverseSexpression(lp.atomsForest[0] as Sexpression);
+        } else {
+            // some kind of structural document error came into play
+            debugger;
+        }
+        this.previousItems.length = 0; // cleanup the tracking mess
+
+        // quick scan any open documents for defun headers. 
+        // Note: This workspace.textDocuments array only ever contains a single document or I would of used it instead.
+        
+        activeDocuments.forEach(doc =>{
+            for (let lnum = 0; lnum < doc.lineCount; lnum++) {
+                const line: vscode.TextLine = doc.lineAt(lnum);
+                const value: string = line.text.toUpperCase();
+                if (value.search(kwRegex) !== -1){                                        
```

Comment:
I think this can not handle the` (defun)` if the line is in comment (specially for the block comment).
I think we need to use the **parser** to parse the whole file, the parser can tokenize the whole lisp file, and can find our interested atoms like (defun) (defun-q)

Reply:
One of the reasons I took this path is because of it being insanely fast. Can I add a LispParser variable to the ReadonlyDocument class so this will be pre-cached for me?